src/QueryFunction.php

```diff
@@ -0,0 +1,128 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * ---------------------------------------------------------------------
+ */
+
+/**
+ *  Query function class
+ **/
+class QueryFunction
+{
+    private string $name;
+
+    private array $params;
+
+    private ?string $alias;
+
+    /**
+     * Create a query expression
+     *
+     * @param string $name Function name
+     * @param array $params Function parameters
+     * @param string|null $alias Function result alias
+     */
+    public function __construct(string $name, array $params = [], ?string $alias = null)
+    {
+        $this->name = $name;
+        $this->params = $params;
+        $this->alias = $alias;
+    }
+
+    /**
+     * Query expression value
+     *
+     * @return string
+     */
+    public function getValue()
+    {
+        global $DB;
+
+        $parameters = $this->params;
+
+        $alias = $this->alias ? ' AS ' . $DB::quoteName($this->alias) : '';
+
+        if ($this->name === 'ADDDATE') {
+            $output = sprintf(
+                'DATE_ADD(%s, INTERVAL %s %s)',
+                $parameters[0], // Date
+                $parameters[1], // Interval
+                strtoupper($parameters[2]) // Unit
+            );
+            $output .= $alias;
+            return $output;
+        }
+
+        return $this->name . '(' . implode(', ', $parameters) . ')' . $alias;
+    }
+
+    public function __toString()
+    {
+        return $this->getValue();
+    }
+
+    /**
+     * Automatically quote a parameter to be used in a SQL function
+     * @param mixed $parameter The parameter to quote
+     * @param bool $identifiers_only If true, regular values will not be quoted
+     * @return mixed
+     */
+    private static function autoQuoteParameter($parameter, bool $identifiers_only = false)
+    {
+        global $DB;
+
+        if ($parameter instanceof self) {
+            return $parameter;
+        }
+        $parameter = trim($parameter);
+        if (preg_match('/^[a-zA-Z0-9_]+\.[a-zA-Z0-9_]+$/', $parameter)) {
+            // quote the table name and column name, then recombine
+            $parts = explode('.', $parameter);
+            $parts = array_map([$DB, 'quoteName'], $parts);
+            $parameter = implode('.', $parts);
+        } else if (!$identifiers_only && !is_numeric($parameter)) {
+            // Quote the parameter if it is not a number
+            $parameter = $DB->quote($parameter);
+        }
+        return $parameter;
+    }
```

Comment:
I'am afraid we will face problem with ambiguous cases in the future if we try to automatically detect wether a parameter is an identifier or a value.
For instance, you may want to use an alias in a function, and, in this case, it will not contain a `.` and will be considered as a value, and will be quoted using simple quotes instead of backticks.

Even if it is more verbose, I would prefer to give responsibility t the caller to define how each parameter have to be handled.
For this, we could use a new `FieldIdentifier` class, that could be used like this: `new QueryFunction('CONCAT', [new FieldIdentifier('glpi_computer.name'), '###', new FieldIdentifier('myalias')])`.

With this, a switch with no possible ambiguity could be used to handle quoting:
```php
switch get_class($param) {
    case FieldIdentifier::class:
    case QueryExpression::class:
        return $param->__toString();
    break;
    default:
        return $DB->quote($parameter);
}
```

Reply:
At that point, does it help any more than expecting the parameters to already be quoted properly before using them with QueryFunction? The only difference is if the developer needs to write `$DB::quoteName($my_field)` or `new FieldIdentifier($my_field)` and if they try using a FieldIdentifier as an array key (like in criteria), they would have to explicitly case it to a string since objects aren't allowed.

Reply:
For now, we could indeed use `$DB::quoteName($my_field)` in constructor call. Maybe it would be usefull in the future to have a `FieldIdentifier` class that could be used by a service that is not using the DB service itself, weh DB will not be global anymore and would be accessible through dependency injection, but it is not for a near future.